Grid selected

thesiv opened the pull request:
Split the wxGrid RANGE_SELECT event into separate SELECTING and SELECTED events.

dkulp committed:
Initial prototype for splitting the wxGrid RANGE_SELECT event into separate SELECTING and SELECTED events
Proposal for #18856

dkulp committed:
Update interface/wx/grid.h

Mark as wxWidgets 3.1.5+

Co-authored-by: VZ <vz-github@zeitlins.org>

dkulp committed:
Update include/wx/generic/grid.h

Comment about when the split occurred

Co-authored-by: VZ <vz-github@zeitlins.org>

dkulp committed:
Fix SELECTED loop

dkulp committed:
Update grid sample to show selecting events

thesiv committed:
Remove the selecting mode flag from wxGridSelection

Remove m_isSelecting and StartSelecting() from wxGridSelection to avoid
inconsistency of wxGridSelection selecting mode state with wxGrid state,
as wxGrid already has m_isDragging field which tells it if the selection
state is final or not.

Instead, just allow wxGrid to specify the event to send from
ExtendCurrentBlock().

We still need a separate EndSelecting() for sending the final
wxEVT_GRID_RANGE_SELECTED event, but send it only for the last selection
block, and not all the selected blocks, as this makes more sense (there
should be one SELECTED event for each block and it was already sent for
the other blocks before) and is consistent with the events generated
when performing the same actions from keyboard.

vadz commented:
@thesiv: I've rebased this branch on the original #2002 and got rid of `WXGRIDSELECT_EVENT_XXX`, it seems unnecessary, and force pushed it. I think we can merge this one while waiting for the other fixes, but I'd still appreciate a comment for the changes of the last commit, as I'm not sure which problem did they fix. TIA!

vadz committed:
Mention that EVT_GRID_RANGE_SELECTED/ING events are new in 3.1.5

Also explained that SELECTED is just the new spelling of SELECT.

vadz committed:
Further improve wxEVT_GRID_RANGE_SELECT{ED,ING} documentation

Explain when it is preferable to handle events of each kind.

vadz committed:
Refactor logging of wxGridRangeSelectEvent in the grid sample

Avoid duplicating the complicated logging code twice, when it's exactly
the same for both the SELECTED and SELECTING events. Just use a common
helper function instead.

thesiv committed:
Return the selecting mode flag back to wxGridSelection

There are some cases when we don't send SELECTING events while dragging:
column/row moving, cells drag handling by the user, start dragging with
Crtl+Click by selected cell. Also SELECTING can be canceled by "slow click".
See wxGrid::DoGridCellLeftUp().

So return m_isSelecting to do not make the code even more complicated.

thesiv commented:
I removed my commit `Remove the selecting mode flag from wxGridSelection` and returned `m_isSelecting`.

vadz commented:
I've created #2028 with the changes that seem to work for me (disregarding the other bugs fixed by your #2024), so I think this one can be closed, but please let me know if you see any problems in my version that didn't exist in this one. TIA!
